include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,402 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <thread>
+#include <memory>
+#include <future>
+
+#ifdef _MSVC_LANG
+#if _MSVC_LANG >= 201703 // _MSVC_LANG == __cplusplus: https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
+#include <execution>
+#endif
+#else
+#if __cplusplus >= 201703
+#include <execution>
+#endif
+#endif
+
+namespace boost { namespace math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class Container>
+void linear_sieve(Z upper_bound, Container &c)
+{
+    size_t least_divisors_size{static_cast<size_t>(upper_bound + 1)};
+    std::unique_ptr<Z[]> least_divisors{new Z[least_divisors_size]{0}};
```

Comment:
Can't this just be `vector<Z>`?

Reply:
It could be, but using a C-style array instead of an `std::vector` yielded lower memory usage and complexity.

Reply:
Hmmm, that doesn't sound right, though, does it?. Like, memory usage could be marginally less with the raw array, but complexity should be identical. What kind of difference are we talking here?

Reply:
It is in micro-optimization territory. I just re-ran it and the difference is 3.72N with the C-style array vs 3.86N for the vector.

Reply:
Are those values consistent between different runs of the benchmark?

Reply:
Although I would generally argue not to leave any performance improvement on the table, I feel like `std::vector` _shouldn't_ be any worse here. At the risk of sounding demanding... have you tried another standard library implementation?

Reply:
They are; benchmark's variation is usually pretty small. These were about +-0.03N.

Reply:
Compiling with Intel C++ and their implementation of the standard library yields the same complexity for the c-style array and the `std::vector`. I guess I would err on the side of more users compiling with gcc/clang and libstdc++ than Intel C++.